apps/src/code-studio/initApp/loadApp.js

```diff
@@ -102,24 +105,35 @@ export function setupApp(appOptions) {
       if (appOptions.level.isProjectLevel && !appOptions.level.edit_blocks) {
         return;
       }
-      // or unless the program is actually the result for a contained level
-      if (!appOptions.hasContainedLevels || appOptions.level.edit_blocks) {
-        if (appOptions.channel && !appOptions.level.edit_blocks) {
-          // Don't send the levelSource or image to Dashboard for channel-backed levels,
-          // unless we are actually editing blocks and not really completing a level
-          // (The levelSource is already stored in the channels API.)
-          delete report.program;
-          delete report.image;
-        } else {
-          // Only locally cache non-channel-backed levels. Use a client-generated
-          // timestamp initially (it will be updated with a timestamp from the server
-          // if we get a response.
-          lastSavedProgram = decodeURIComponent(report.program);
-          clientState.writeSourceForLevel(appOptions.scriptName, appOptions.serverLevelId, +new Date(), lastSavedProgram);
-        }
+      if (appOptions.channel && !appOptions.level.edit_blocks &&
+          !appOptions.hasContainedLevels) {
+        // Unless we are actually editing blocks and not really completing a
+        // level, or if this is a contained level, don't send the levelSource or
+        // image to Dashboard for channel-backed levels (The levelSource is
+        // already stored in the channels API.)
+        delete report.program;
+        delete report.image;
+      } else {
+        // Only locally cache non-channel-backed levels. Use a client-generated
+        // timestamp initially (it will be updated with a timestamp from the server
+        // if we get a response.
+        lastSavedProgram = decodeURIComponent(report.program);
+
+        // If the program is the result for a contained level, store it with
+        // the contained level id
+        const levelId = appOptions.hasContainedLevels ?
+          getContainedLevelId() :
+          appOptions.serverLevelId;
+        clientState.writeSourceForLevel(appOptions.scriptName, levelId,
+            +new Date(), lastSavedProgram);
+      }
+      // report.callback will already have the correct milestone post URL in
+      // the contained level case
```

Comment:
`report.callback` is only set by the caller in the contained level case, otherwise it's undefined here. It might be clearer to treat `appOptions.report.callback` and `report.callback` as a default and an override, e.g.; https://github.com/code-dot-org/code-dot-org/compare/contained-offline-override?expand=1#diff-c3dded2308a12c2c9b01a07d1eeb3db3R133

appOptions.report.callback will have a URL with the outer level's level_id, but with a contained level it needs to be the inner level's level_id.